contracts/collection.scilla

```diff
@@ -0,0 +1,575 @@
+    RequireNotPaused;
+
+    maybe_collection_id <- token_collection[token_address][token_id];
+    
+    match maybe_collection_id with 
+    | Some maybe_collection_id => 
+        delete token_collection[token_address][token_id];
+        e = {
+            _eventname : "RemovedTokenFromCollection";
+            token_address: token_address;
+            token_id: token_id;
+            collection_id: collection_id
+          };
+        event e
+        
+    | None =>
+        error = TokenDoesNotExistInCollection;
+        Throw error
+    end;
+
+    (* Deleting the corresponding record in has_had_first_sale if exist *)
+    is_sale_record <- exists has_had_first_sale[token_address][token_id];
+
+    match is_sale_record with 
+    | True  =>
+      delete has_had_first_sale[token_address][token_id]
+    | False => (* No operation *)
+    end
+end
+
+
+(* Used by a brand_owner to request an NFT to their collection *)
+transition RequestTokenToCollection(
+    token_address: ByStr20 with contract 
+        field token_owners: Map Uint256 ByStr20 
+    end,
+    token_id: Uint256, 
+    collection_id: Uint32
+    )
+    RequireBrandOwner collection_id;
+    RequireTokenNotInCollection token_address token_id;
+    RequireNotPaused;
+
+    (* When the same brand_owner uses this function a second time we just overwrite *)
+    request = Requested collection_id;
+    requests[token_address][token_id][_sender] := request;
+    commission_fee <- collection_commission_fee[collection_id];
+
+    e = {
+        _eventname : "RequestTokenToCollection";
+        token_address: token_address;
+        token_id: token_id;
+        collection_id: collection_id;
+        commission_fee: commission_fee
+    };
+    event e
+end
+
+(* Lets a brand_owner delete a request if one exist. *)
+transition DeleteRequestTokenToCollection(
+    token_address: ByStr20 with contract 
+        field token_owners: Map Uint256 ByStr20 
+    end,
+    token_id: Uint256, 
+    collection_id: Uint32
+    )
+    RequireBrandOwner collection_id;
+    RequireNotPaused;
+    
+    maybe_request <- requests[token_address][token_id][_sender];
+    
+    match maybe_request with
+      | Some maybe_request => 
+      delete requests[token_address][token_id][_sender];
+
+      e = {
+        _eventname : "RequestTokenToCollectionDeleted";
+        token_address: token_address;
+        token_id: token_id;
+        collection_id: collection_id
+      };
+      event e
+
+      | None => 
+      error = RequestDoesNotExist;
+      Throw error
+    end
+end
+
+(* Used by a NFT owner to accept a request to have one of their NFTs added to a collection *)
+transition AcceptCollectionRequest(
+    token_address: ByStr20 with contract 
+        field token_owners: Map Uint256 ByStr20 
+    end,
+    token_id: Uint256, 
+    brand_address: ByStr20
+    )
+    RequireTokenOwner token_address token_id;
+    RequireTokenNotInCollection token_address token_id;
+    RequireAllowedUser _sender;
```

Comment:
@sam-zilliqa do we need an explicit check ensuring that token is not in any of the fixed price contracts? I understand that this check is sufficient for making sure that token is not listed in the auction contract.